aptos-move/framework/aptos-framework/sources/coin.move

```diff
@@ -150,6 +193,252 @@ module aptos_framework::coin {
+        let metadata_addr = object::object_address(&metadata);
+        if (exists<PairedCoinType>(metadata_addr)) {
+            option::some(borrow_global<PairedCoinType>(metadata_addr).type)
+        } else {
+            option::none()
+        }
+    }
+
+    /// Conversion from coin to fungible asset
+    public fun coin_to_fungible_asset<CoinType>(
+        coin: Coin<CoinType>
+    ): FungibleAsset acquires CoinConversionMap, CoinInfo {
+        let metadata = ensure_paired_metadata<CoinType>();
+        let amount = burn_internal(coin);
+        fungible_asset::mint_internal(metadata, amount)
+    }
+
+    /// Conversion from fungible asset to coin. Not public to push the migration to FA.
+    fun fungible_asset_to_coin<CoinType>(
+        fungible_asset: FungibleAsset
+    ): Coin<CoinType> acquires CoinInfo, PairedCoinType {
+        let metadata_addr = object::object_address(&fungible_asset::metadata_from_asset(&fungible_asset));
+        assert!(
+            object::object_exists<PairedCoinType>(metadata_addr),
+            error::not_found(EPAIRED_COIN)
+        );
+        let coin_type_info = borrow_global<PairedCoinType>(metadata_addr).type;
+        assert!(coin_type_info == type_info::type_of<CoinType>(), error::invalid_argument(ECOIN_TYPE_MISMATCH));
+        let amount = fungible_asset::burn_internal(fungible_asset);
+        mint_internal<CoinType>(amount)
+    }
+
+    inline fun assert_paired_metadata_exists<CoinType>(): Object<Metadata> {
+        let metadata_opt = paired_metadata<CoinType>();
+        assert!(option::is_some(&metadata_opt), error::not_found(EPAIRED_FUNGIBLE_ASSET));
+        option::destroy_some(metadata_opt)
+    }
+
+    #[view]
+    /// Check whether `MintRef` has not been taken.
+    public fun paired_mint_ref_exists<CoinType>(): bool acquires CoinConversionMap, PairedFungibleAssetRefs {
+        let metadata = assert_paired_metadata_exists<CoinType>();
+        let metadata_addr = object_address(&metadata);
+        assert!(exists<PairedFungibleAssetRefs>(metadata_addr), error::internal(EPAIRED_FUNGIBLE_ASSET_REFS_NOT_FOUND));
+        option::is_some(&borrow_global<PairedFungibleAssetRefs>(metadata_addr).mint_ref_opt)
+    }
+
+    /// Get the `MintRef` of paired fungible asset of a coin type from `MintCapability`.
+    public fun get_paired_mint_ref<CoinType>(
+        _: &MintCapability<CoinType>
+    ): (MintRef, MintRefReceipt) acquires CoinConversionMap, PairedFungibleAssetRefs {
+        let metadata = assert_paired_metadata_exists<CoinType>();
+        let metadata_addr = object_address(&metadata);
+        assert!(exists<PairedFungibleAssetRefs>(metadata_addr), error::internal(EPAIRED_FUNGIBLE_ASSET_REFS_NOT_FOUND));
+        let mint_ref_opt = &mut borrow_global_mut<PairedFungibleAssetRefs>(metadata_addr).mint_ref_opt;
+        assert!(option::is_some(mint_ref_opt), error::not_found(EMINT_REF_NOT_FOUND));
+        (option::extract(mint_ref_opt), MintRefReceipt { metadata })
```

Comment:
I don't think you need an option in this case, since we require to return the MintRef at the end and they are droppable, right?

We can make a copy and return and drop it with the receipt

Reply:
refs are not copyable